### ISSUE TITLE
Path module dirname path separator '/' inconsistent under windows

### ISSUE THREAD

**Issue by visualjazz-skinread** [joyent/node#16527]
Have found under windows that path.dirname is returning a string with the path separator `/` while path.sep is returning `\`. Unsure if there this could be an underlying config issue, however it seems to potentially only affect path.dirname as I've found with some test cases.

example
```
path.dirname(file); // value is "src/assets"
path.sep; // value is  "\"
path.dirname(file).split(path.sep); // array is full path [ "src/assets" ] 
```

Work around I have found is wrapping the path.dirname in path.normalize fixes the mismatch
```
path.normalize(path.dirname(file)); // "src\assets"
```

Windows 8 64-bit
Node 0.12.0

**Comment by cjihrig** [comment]
I think the problem is that your `file` variable uses Unix slashes instead of Windows backslashes. Based on your linked commit, it appears that applying `path.normalize()` solves the problem.

**Comment by visualjazz-skinread** [comment]
Would it make any sense to normalize the value returned from `path.normalize`? - it seems that mismatches could be common in real world configurations and you would always want the separators normalized regardless.

Also `\` is not documented as a valid value for `path.sep` (https://nodejs.org/api/path.html#path_path_sep) so I found it confusing that it was possible to arrive at this outcome.
Cheers

**Comment by cjihrig** [comment]
`\` is a valid value for `path.sep`. The docs and REPL show `'\\'`, which is escaped for a string literal. This actually corresponds to the `\` character, which is what is actually used on Windows.

As for whether or not, `path.dirname()` should normalize the return value - I don't have a strong preference. @joyent/node-coreteam any opinions?

**Comment by misterdjules** [comment]
Thinking out loud, as I'm not a user of the `path` module on multiple platforms.

It seems that `path.dirname` and `path.basename` correctly handle Windows and UNIX paths transparently on Windows, but don't transparently handle Windows path on UNIX. The reason is that on Windows, forward-slashes are not valid characters in file paths, whereas backslashes are valid characters in UNIX paths.

In other words, while normalizing the output of `path.dirname` might look like a solution on Windows to transparently process UNIX paths, it seems it wouldn't fix the problem of transparently processing Windows paths on UNIX.

I'm using the term "transparently" often because if we know in advance the way we want to process these paths, we can always use the platform specific `path.posix` and `path.win32` APIs.

Let's imagine that `path.dirname` and `path.basename` handle Windows paths transparently and correctly on UNIX. I would think that `path.dirname` would return the actual directory name of the path that was passed as input, not mutate silently the directory name of a path to suit the conventions of my current platform.

Instead, I would prefer to have a `path.split` function that would take care of transparently handling paths from any platform on any platform, but I'm wondering if that's possible at all.

In other words:
1. I don't think we want to mutate the results of `path.dirname` by default.
2. I'm not sure there's a way to handle UNIX and Windows paths transparently when running on UNIX.
3. If you know the platform of the paths that you get on input, you can use platform specific `path.posix` or `path.win32` APIs, or you can normalize them.

Thoughts?

**Comment by cjihrig** [comment]
@misterdjules Given your comments, I think it's best to not normalize the results. Therefore, I'm going to close this issue.
